flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecMultiJoin.java

```diff
@@ -0,0 +1,444 @@
+            @JsonProperty(FIELD_NAME_OUTPUT_TYPE) RowType outputType,
+            @JsonProperty(FIELD_NAME_DESCRIPTION) String description,
+            @JsonProperty(FIELD_NAME_KEY_SELECTORS)
+                    List<KeySelector<RowData, RowData>> commonJoinKeySelectorsForTransformation,
+            @JsonProperty(FIELD_NAME_COMMON_KEY_TYPE) InternalTypeInfo<RowData> commonJoinKeyType) {
+        super(id, context, persistedConfig, inputProperties, outputType, description);
+        validateInputs(inputProperties, joinTypes, joinConditions, inputSideSpecs);
+
+        this.joinTypes = checkNotNull(joinTypes);
+        this.joinConditions = checkNotNull(joinConditions);
+        this.joinAttributeMap = checkNotNull(joinAttributeMap);
+        this.inputSideSpecs = checkNotNull(inputSideSpecs);
+        this.stateMetadataList = stateMetadataList;
+        this.commonJoinKeySelectorsForTransformation =
+                checkNotNull(commonJoinKeySelectorsForTransformation);
+        this.commonJoinKeyType = checkNotNull(commonJoinKeyType);
+    }
+
+    private void validateInputs(
+            List<InputProperty> inputProperties,
+            List<JoinType> joinTypes,
+            List<? extends @Nullable RexNode> joinConditions,
+            List<JoinInputSideSpec> inputSideSpecs) {
+        checkArgument(
+                inputProperties.size() >= 2, "Multi-input join operator needs at least 2 inputs.");
+        checkArgument(
+                joinTypes.size() == inputProperties.size(),
+                "Size of joinTypes must match the number of inputs.");
+        checkArgument(
+                joinConditions.size() == inputProperties.size(),
+                "Size of joinConditions must match the number of inputs.");
+        checkArgument(
+                inputSideSpecs.size() == inputProperties.size(),
+                "Size of inputSideSpecs must match the number of inputs.");
+    }
+
+    private static String[] generateStateNames(int numInputs) {
+        return IntStream.range(0, numInputs)
+                .mapToObj(i -> "inputState-" + i)
+                .toArray(String[]::new);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected Transformation<RowData> translateToPlanInternal(
+            PlannerBase planner, ExecNodeConfig config) {
+        final List<ExecEdge> inputEdges = getInputEdges();
+        final int numInputs = inputEdges.size();
+        final ClassLoader classLoader = planner.getFlinkContext().getClassLoader();
+
+        // Prepare input transformations and types
+        final List<Transformation<RowData>> inputTransforms = new ArrayList<>(numInputs);
+        final List<InternalTypeInfo<RowData>> inputTypeInfos = new ArrayList<>(numInputs);
+        final List<RowType> inputRowTypes = new ArrayList<>(numInputs);
+
+        for (ExecEdge inputEdge : inputEdges) {
+            Transformation<RowData> transform =
```

Comment:
nit: add `final`. it takes some time to get into this habbit but it really happens to check that variable is assigned exactly one time and is not null accidentally.